REGRESSION (281966@main): Unable to drag images from Photos into Mail compose on macOS

whsieh opened the pull request:
#### 59cdc6c756949563050436d2e83326afa7c93840
<pre>
REGRESSION (281966@main): Unable to drag images from Photos into Mail compose on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=278749">https://bugs.webkit.org/show_bug.cgi?id=278749</a>
<a href="https://rdar.apple.com/134780008">rdar://134780008</a>

Reviewed by Aditya Keerthi and Charlie Wolfe.

The refactoring around `WebViewImpl::performDragOperation` in <a href="https://commits.webkit.org/281966@main">https://commits.webkit.org/281966@main</a>
introduced a bug, wherein dragging multiple promised files over a web view causes the drop to fail.
This is because the reference to `page` is moved (and cleared) when invoking the iteration block,
causing subsequent calls to exit early (in the `!page` branch).

Fix this by copying a reference to the `WebPageProxy` into each iteration block.

Tested by an existing API test, after augmenting test infrastructure (see below).

Test: WKAttachmentTestsMac.InsertDroppedFilePromisesAsAttachments

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::handleLegacyFilesPasteboard):
* Tools/TestWebKitAPI/mac/TestDraggingInfo.mm:

Refactor some test infrastructure that was intended to simulate dragging and dropping one or more
promised file(s), such that it&apos;s consistent with how AppKit delivers promised data on drop.
Currently, the test harness handles multiple promised file URLs (written via `-writePromisedFiles:`)
by creating a single drag item that contains multiple URLs. However, when dragging and dropping
multiple files or photos into Mail compose, it&apos;s actually dropped as multiple drag items, each with
a single URL and type.

(-[TestDraggingInfo enumerateDraggingItemsWithOptions:forView:classes:searchOptions:usingBlock:]):
* Tools/TestWebKitAPI/mac/TestFilePromiseReceiver.h:
* Tools/TestWebKitAPI/mac/TestFilePromiseReceiver.mm:
(-[TestFilePromiseReceiver initWithTypeIdentifier:fileURL:]):
(-[TestFilePromiseReceiver fileTypes]):
(-[TestFilePromiseReceiver fileNames]):
(-[TestFilePromiseReceiver dealloc]):
(-[TestFilePromiseReceiver receivePromisedFilesAtDestination:options:operationQueue:reader:]):
(-[TestFilePromiseReceiver initWithPromisedTypeIdentifiers:dragAndDropSimulator:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/282828@main">https://commits.webkit.org/282828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dba4acae93565b6eddf80ff82eb48713fd0e2c5c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64410 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43774 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17006 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68431 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15017 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66529 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15297 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51827 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10358 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67478 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40463 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55732 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32446 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/37133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13111 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13891 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13440 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70132 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8357 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12953 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59152 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8391 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55822 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59319 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14210 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6907 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/588 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39588 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/40666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40409 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/mac/TestDraggingInfo.mm:
```diff
@@ -100,23 +100,26 @@ - (void)enumerateDraggingItemsWithOptions:(NSDraggingItemEnumerationOptions)enum
+
+        NSUInteger itemIndex = 0;
+        for (NSURL *promisedFileURL in externalPromisedFiles.get()) {
+            auto receiver = adoptNS([[TestFilePromiseReceiver alloc] initWithTypeIdentifier:[promisedTypeIdentifiers objectAtIndex:itemIndex] fileURL:promisedFileURL]);
```
RetainPtr instead of auto?

charliewolfe commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4123,11 +4123,11 @@ static bool handleLegacyFilesPasteboard(id<NSDraggingInfo> draggingInfo, Box<Web
             if (errorOrNil)
                 return;
 
-            RunLoop::main().dispatch([page = WTFMove(page), path = RetainPtr { fileURL.path }, fileNames, fileCount, dragData, pasteboardName] () mutable {
```
Why is this move wrong? Don’t you just need to remove the first one?

whsieh replied:
Good catch — restored this part.

charliewolfe commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4123,11 +4123,11 @@ static bool handleLegacyFilesPasteboard(id<NSDraggingInfo> draggingInfo, Box<Web
     [draggingInfo enumerateDraggingItemsWithOptions:0 forView:view.autorelease() classes:@[NSFilePromiseReceiver.class] searchOptions:@{ } usingBlock:[&](NSDraggingItem *draggingItem, NSInteger idx, BOOL *stop) {
         auto queue = adoptNS([NSOperationQueue new]);
-        [draggingItem.item receivePromisedFilesAtDestination:dropDestination options:@{ } operationQueue:queue.get() reader:[page = WTFMove(page), fileNames, fileCount, dragData, pasteboardName](NSURL *fileURL, NSError *errorOrNil) {
+        [draggingItem.item receivePromisedFilesAtDestination:dropDestination options:@{ } operationQueue:queue.get() reader:[page, fileNames, fileCount, dragData, pasteboardName](NSURL *fileURL, NSError *errorOrNil) {
```
I think the `RefPtr` parameter in this function is incorrect - `page` can’t be null, and smart pointers shouldn’t be passed as parameters by value. The parameter should change to `WebPageProxy&` and this line should be `page = Ref { page }`.

whsieh replied:
That's true — I can make this a `Ref` instead.

(I'll need to update the signature of `performDragWithLegacyFiles` as well)

whsieh committed:
REGRESSION (281966@main): Unable to drag images from Photos into Mail compose on macOS
https://bugs.webkit.org/show_bug.cgi?id=278749
rdar://134780008

Reviewed by Aditya Keerthi and Charlie Wolfe.

The refactoring around `WebViewImpl::performDragOperation` in https://commits.webkit.org/281966@main
introduced a bug, wherein dragging multiple promised files over a web view causes the drop to fail.
This is because the reference to `page` is moved (and cleared) when invoking the iteration block,
causing subsequent calls to exit early (in the `!page` branch).

Fix this by copying a reference to the `WebPageProxy` into each iteration block.

Tested by an existing API test, after augmenting test infrastructure (see below).

Test: WKAttachmentTestsMac.InsertDroppedFilePromisesAsAttachments

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::handleLegacyFilesPasteboard):
* Tools/TestWebKitAPI/mac/TestDraggingInfo.mm:

Refactor some test infrastructure that was intended to simulate dragging and dropping one or more
promised file(s), such that it's consistent with how AppKit delivers promised data on drop.
Currently, the test harness handles multiple promised file URLs (written via `-writePromisedFiles:`)
by creating a single drag item that contains multiple URLs. However, when dragging and dropping
multiple files or photos into Mail compose, it's actually dropped as multiple drag items, each with
a single URL and type.

(-[TestDraggingInfo enumerateDraggingItemsWithOptions:forView:classes:searchOptions:usingBlock:]):
* Tools/TestWebKitAPI/mac/TestFilePromiseReceiver.h:
* Tools/TestWebKitAPI/mac/TestFilePromiseReceiver.mm:
(-[TestFilePromiseReceiver initWithTypeIdentifier:fileURL:]):
(-[TestFilePromiseReceiver fileTypes]):
(-[TestFilePromiseReceiver fileNames]):
(-[TestFilePromiseReceiver dealloc]):
(-[TestFilePromiseReceiver receivePromisedFilesAtDestination:options:operationQueue:reader:]):
(-[TestFilePromiseReceiver initWithPromisedTypeIdentifiers:dragAndDropSimulator:]): Deleted.

Canonical link: https://commits.webkit.org/282828@main
